astropy/utils/decorators.py

```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    object::
+
+        >>> @format_doc(add)
+        ... def another_add(num1, num2):
+        ...     return num1 + num2
+        ...
+        >>> help(another_add) # doctest: +SKIP
+        Help on function another_add in module __main__:
+        <BLANKLINE>
+        another_add(num1, num2)
+            Perform addition.
+            Parameters
+            ----------
+            num1, num2 : Numbers
+            Returns
+            -------
+            result: Number
+                result of num1 + num2
+
+    But be aware that this decorator *only* formats the given docstring not
+    the strings passed as ``args`` or ``kwargs`` (not even the original
+    docstring)::
+
+        >>> @format_doc(doc, 'addition', op='+')
+        ... def yet_another_add(num1, num2):
+        ...    '''This one is good for {0}.'''
+        ...    return num1 + num2
+        ...
+        >>> help(yet_another_add) # doctest: +SKIP
+        Help on function yet_another_add in module __main__:
+        <BLANKLINE>
+        yet_another_add(num1, num2)
+            Perform addition.
+            Parameters
+            ----------
+            num1, num2 : Numbers
+            Returns
+            -------
+            result: Number
+                result of num1 + num2
+            This one is good for {0}.
+
+    To work around it you could specify the docstring to be ``'__doc__'``::
+
+        >>> @format_doc('__doc__', 'addition')
+        ... def last_add_i_swear(num1, num2):
+        ...    '''This one is good for {0}.'''
+        ...    return num1 + num2
+        ...
+        >>> help(last_add_i_swear) # doctest: +SKIP
+        Help on function last_add_i_swear in module __main__:
+        <BLANKLINE>
+        last_add_i_swear(num1, num2)
+            This one is good for addition.
+
+    Using it with ``'__doc__'`` as docstring allows to use the decorator twice
+    on an object to first parse the new docstring and then to parse the
+    original docstring or the ``args`` and ``kwargs``.
+    """
+    def set_docstring(func):
+        # Not a string so assume we want the saved doc from the object
+        if not isinstance(docstring, six.string_types):
+            doc = docstring.__doc__
+        elif docstring != '__doc__':
+            doc = docstring
+        else:
+            doc = func.__doc__
+            # Delete documentation in this case so we don't end up with
+            # awkwardly self-inserted docs.
+            func.__doc__ = None
+
+        if not doc:
+            # In case the docstring is empty it's probably not what was wanted.
+            raise ValueError('docstring must be a string or containing a '
+                             'docstring that is not empty.')
+
+        # If the original has a not-empty docstring append it to the format
+        # kwargs.
+        kwargs['__doc__'] = func.__doc__ or ''
+        func.__doc__ = doc.format(*args, **kwargs)
```

Comment:
To support classes on Python 2 we might do the following:
```python
if six.PY2 and isinstance(obj, type):
    obj = type(obj.__name__, (obj,), {'__doc__': doc.format(*args, **kwargs),
                                                         '__module__': obj.__module__})
else:
    obj.__doc__ = doc.format(*args, **kwargs)
```

It's unfortunate (and at times confusing) to insert this trivial subclass just to change the `__doc__`, but it'll get the job done for now, and probably won't be used much anyways.  But that way you can get rid of all the provisos.